[DataGrid] Add Density selector (#606)
mui-org/material-ui-x: docs/src/pages/components/data-grid/rendering/rendering.md

{{"demo": "pages/components/data-grid/rendering/AutoHeightGrid.js", "bg": "inline"}}

> ## Size

Reviewer: How about we use the two wordings at the same time so developers, when searching, e.g. with Algolia can find the content?

```suggestion
## Size/Density
```

Author: done

You can change the density of the rows and the column header.

### Density prop

> The vertical density of the data grid can be set using the `density` prop. Regardless of its setting, the `density` prop will respect values provided using the `rowHeight` or `headerHeight` props.

Reviewer 2: ```suggestion
The vertical density of the data grid can be set using the `density` prop. The `density` prop applies the values determined by the `rowHeight` and `headerHeight` props, if supplied.  The user can override this setting with the toolbar density selector, if provided. 
```

(Hopefully I got the meaning correct this time,  now that I better understand it. Comment if it still isn't correct, or isn't clear. 🙂)

{{"demo": "pages/components/data-grid/rendering/SizePickerSmallGrid.js", "bg": "inline"}}

> ### Size picker button

Reviewer: There is an issue with the experimental feature. The documented demo doesn't work. The toolbar is hidden by default. This is related to https://github.com/mui-org/material-ui-x/pull/606#issuecomment-737130908.

Author: Ok, yes I forgot. I enabled the toolbar for the demo

> There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.

Reviewer: use the generic to not exclude XGrid.

```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the data grid is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

Reviewer 2: ```suggestion
The vertical density data grid  can be set using the `size` prop. Regardless of its setting, the `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

Avoid mixing size and density terminology, avoid duplicating info from the API docs (defaults, prop values). (I don't hold a strong opinion on this second point.

> ### Density selector button

Reviewer 2: ```suggestion
### Density selector
```

> There are three possible density options to choose from: `small`, `medium` and `large`. By default the density of the `DataGrid` is set to `medium`. The default grid density can be changed by setting the `density` propery. The `density` propery works with relation to the `rowHeight` and `headerHeight` properties. This means that it will respect the values provided using `rowHeight` or `headerHeight` properties.

Reviewer: ```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by setting the `density` prop. The `density` prop works with relation to the `rowHeight` and `headerHeight` properties. This means that it will respect the values provided using `rowHeight` or `headerHeight` props.
```

Reviewer 2: Based off @oliviertassinari's:
```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by applying the `density` prop. The `density` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

> The default way for a user to change the density of the data grid is by using the density selector component from the toolbar. In addition, there is a prop to hide the density selector.

Reviewer 2: ```suggestion
The user can change the density of the data grid  by using the density selector from the toolbar. This can be hidden with the `disableDensitySelector` prop.
```

Reviewer 2: Should this be called `hideDensitySelector` to better reflect it's function?

![image](https://user-images.githubusercontent.com/357702/101086127-4f87a680-35a8-11eb-96e2-4b478e7b8e92.png)

Reviewer: IMHO, it would be better to drop these props and move to the opt-in toolbar + composable API, one of the last few points of #633 that isn't checked yet.

Author: We can remove them later. For now the name matches the name of the other `toolbar` props.

> ### Density Picker button

Reviewer: Point 2. https://trello.com/c/1Q7hqXHu/2714-technical-documentation-guidelines

```suggestion
### Density picker button
```

> The default way to change the density of the data grid is by using the `DensityPicker` component from the toolbar. In addition, there is a prop to hide the `DensityPicker`.

Reviewer: This component is private, developers shouldn't know about it.

```suggestion
The default way to change the density of the data grid is by using the density picker component from the toolbar. In addition, there is a prop to hide the density picker.
```

Reviewer 2: ```suggestion
The default way for a user to change the density of the data grid is by using the density picker component from the toolbar. In addition, there is a prop to hide the density picker.
```

> The default way to change the density of the `DataGrid` is by using the `DensityPicker` component from the toolbar. In addition there is a property to hide the `DensityPicker` from the toolbar.

Reviewer: Components are in a global name space, `DensityPicker` is not OK, should be `GridDensityPicker`.

> The default way to change the density of the `DataGrid` is by using the `DensityPicker` component from the toolbar. In addition there is a property to hide the `DensityPicker` from the toolbar.

Reviewer 2: ```suggestion
The default way to change the density of the data grid is by using the `DensityPicker` component on the toolbar. The `disableDensityPicker` prop can be used to hide this control.
```

### Size prop

> There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the data grid is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.

Reviewer 2: That's odd, I left a comment with a suggestion that fixed some of the issues with this description, while leaving some of it open for discussion. Could we please not just overrule in those cases?

Reviewer 2: Thanks for adding the suggestion back @oliviertassinari. The comment was something to the effect that it isn't necessary to reproduce the API documentation in the description. This change also resolves the potential confusion around what the options actually apply to, and also the use of "default" to mean two different things.

Author: I'm sorry about that @mbrookes I did merge it. I think git merges it weirdly if there are 2 suggestions on the same block.

Reviewer 2: Good to know, I'll have to keep an eye out for that, thanks.
